File: vortex/backend/src/main/java/com/vortex/service/IngestionService.java
package com.vortex.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.tika.Tika;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class IngestionService {

    private final LiteLlmService liteLlmService;
    private final ChromaService chromaService;
    private final Tika tika = new Tika();

    public void ingestDocument(MultipartFile file, String apiKey, int chunkSize, int chunkOverlap) throws Exception {
        String filename = file.getOriginalFilename();
        log.info("Starting intelligent ingestion for file: {}", filename);

        // 1. Extract Text
        String content = tika.parseToString(file.getInputStream());
        if (content == null || content.trim().isEmpty()) return;

        // 2. Determine Semantic Topic (AI-powered metadata)
        String sample = content.substring(0, Math.min(content.length(), 2000));
        String topic = liteLlmService.getCompletion(
                "Describe this document in 5 words or less: " + sample, apiKey).block();
        log.info("Semantic Topic Identified: {}", topic);

        // 3. De-duplicate: Clean existing records for this file
        chromaService.deleteByFilename(filename);

        // 4. Advanced Chunking
        List<String> chunks = splitIntoChunks(content, chunkSize, chunkOverlap);

        // 5. Embed and Store with Topic Metadata
        for (int i = 0; i < chunks.size(); i++) {
            String chunk = chunks.get(i);
            List<Double> embedding = liteLlmService.getEmbedding(chunk, apiKey).block();
            if (embedding != null) {
                chromaService.upsert(chunk, embedding, filename, i, topic);
            }
        }
        
        log.info("Successfully indexed {} chunks for: {} (Topic: {})", chunks.size(), filename, topic);
    }

    private List<String> splitIntoChunks(String text, int size, int overlap) {
        List<String> chunks = new ArrayList<>();
        int start = 0;
        while (start < text.length()) {
            int end = Math.min(start + size, text.length());
            chunks.add(text.substring(start, end));
            start += (size - overlap);
            if (start >= text.length() || size <= overlap) break;
        }
        return chunks;
    }
}

File: vortex/backend-py/services/ingestion_service.py
from tika import parser
import logging
import litellm
import os
from typing import List
from services.chroma_service import ChromaService

logger = logging.getLogger(__name__)

class IngestionService:
    def __init__(self, chroma_service: ChromaService):
        self.chroma_service = chroma_service
        self.proxy_base_url = os.getenv("LITELLM_PROXY_BASE_URL")

    async def ingest_document(self, file_content: bytes, filename: str, api_key: str, chunk_size: int = 1000, chunk_overlap: int = 200):
        logger.info(f"Ingesting: {filename}")
        
        parsed = parser.from_buffer(file_content)
        content = parsed.get("content", "").strip()
        if not content: return

        # 1. ENHANCED: Semantic Topic Extraction
        # We use the LLM to 'look' at the first 2000 chars to determine the topic
        topic = await self._determine_topic(content[:2000], api_key)
        logger.info(f"Semantic Topic Identified: {topic}")

        # 2. De-duplicate
        self.chroma_service.delete_by_filename(filename)

        # 3. Chunking
        chunks = self._split_into_chunks(content, chunk_size, chunk_overlap)
        
        # 4. Embed and Store
        for i, chunk in enumerate(chunks):
            try:
                embedding = self.chroma_service._get_embedding(chunk, api_key)
                self.chroma_service.upsert(chunk, embedding, filename, i, topic)
            except Exception as e:
                logger.error(f"Chunk {i} failed: {e}")

        logger.info(f"Indexed {len(chunks)} chunks for: {filename}")

    async def _determine_topic(self, sample_text: str, api_key: str) -> str:
        """Uses LLM to describe the content regardless of filename."""
        try:
            response = litellm.completion(
                model="gemini-1.5-flash", # Use a fast model for tagging
                messages=[{"role": "user", "content": f"Describe this document content in 5 words or less. Be descriptive. Content:\n{sample_text}"}],
                api_base=self.proxy_base_url,
                api_key=api_key
            )
            return response.choices[0].message.content.strip()
        except Exception as e:
            logger.warning(f"Topic extraction failed: {e}")
            return "General Knowledge"

    def _split_into_chunks(self, text: str, size: int, overlap: int) -> List[str]:
        chunks = []
        start = 0
        while start < len(text):
            end = min(start + size, len(text))
            chunks.append(text[start:end])
            start += (size - overlap)
            if start >= len(text) or size <= overlap: break
        return chunks